src/mongo/client/dbclient.cpp

```diff
@@ -1066,15 +1070,175 @@ namespace mongo {
+                                                  saslCommandMechanismFieldName << "SCRAM-SHA-1");
+
+            BSONObj saslFollowupCommandPrefix = BSON("copydb" << 1 <<
+                                                     "fromhost" << fromhost <<
+                                                     "fromdb" << fromdb <<
+                                                     "todb" << todb);
+
+            BSONObj saslCommandPrefix = saslFirstCommandPrefix;
+            BSONObj inputObj = BSON(saslCommandPayloadFieldName << "");
+            bool isServerDone = false;
+
+            // send copydbsaslstart, then continue to send copydb until we are done.
+            while (!session->isDone()) {
+                std::string payload;
+                BSONType type;
+
+                Status status = saslExtractPayload(inputObj, &payload, &type);
+                if (!status.isOK()) {
+                    throw DBException( str::stream() << "sasl session failure: " << status.reason(), 14355 );
+                }
+
+                std::string responsePayload;
+                status = session->step(payload, &responsePayload);
+                if (!status.isOK()) {
+                    throw DBException( str::stream() << "sasl session failure: " << status.reason(), 14356 );
+                }
+
+                // build command to send to server
+                BSONObjBuilder commandBuilder;
+                commandBuilder.appendElements(saslCommandPrefix);
+                commandBuilder.appendBinData(saslCommandPayloadFieldName,
+                                             int(responsePayload.size()),
+                                             BinDataGeneral,
+                                             responsePayload.c_str());
+                BSONElement conversationId = inputObj[saslCommandConversationIdFieldName];
+                if (!conversationId.eoo()) {
+                    commandBuilder.append(conversationId);
+                }
+
+                BSONObj command = commandBuilder.obj();
+                bool ok = runCommand("admin", command, inputObj);
+
+                ErrorCodes::Error code =
+                    ErrorCodes::fromInt(inputObj[saslCommandCodeFieldName].numberInt());
+
+                if (!ok || code != ErrorCodes::OK) {
+                    if (code == ErrorCodes::OK)
+                        code = ErrorCodes::UnknownError;
+                    // TODO handle
+                    // attempt to give a sane error message
+                    if ( inputObj.hasField("errmsg") ) {
+                        BSONElement e = inputObj.getField("errmsg");
+                        if (e.type() == String &&
+                            strstr( e.valuestr(), "no such cmd: saslStart")) {
+                            throw UserException( 13423,
+                                                 "fromhost doesn't support SCRAM-SHA-1, use MONGODB-CR");
+                        }
```

Comment:
then we can't give a nice explanation and we'll have to just throw an operation exception instead